### PR TITLE
Adding test for OCP OLM based operator and CR migration

### PR DESCRIPTION
**Pull request by gejain**
Adding test for OCP OLM based operator and CR migration
- Postgresql Operator
- Elasticsearch Operator

**Which issue(s) this PR fixes** (optional)
Closes #PTX-24434, PTX-24435

**Special notes for your reviewer**:
Passed Jobs -
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Geet/job/TorpedoOcpTest/4
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Geet/job/TorpedoOcpTest/5

### CODE REVIEW

File: tests/basic/async_dr_test.go [hunk 1]
```diff
@@ -817,10 +985,253 @@ func hardSetConfig(configPath string) error {
...
+		log.Infof("Error unmarshalling JSON: %v", err)
+		return 0, err
+	}
+	return len(appReg.Resources), nil
+}
+
+func patchStashStrategy(crName string) error {
+	var patches []PatchOperation
+	numResources, err := getNumResources(crName)
+	log.FailOnError(err, "Failed to get number of resources")
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1206]: return err?

Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

**Review by Rohit-PX — APPROVED** [review]
LGTM.

File: tests/basic/async_dr_test.go [hunk 2]
```diff
@@ -476,9 +481,172 @@ var _ = Describe("{StorkctlPerformFailoverFailbackDefaultMetroMultiple}", func()
...
+	BeforeEach(func() {
+		if !kubeConfigWritten {
+			WriteKubeconfigToFiles()
+			kubeConfigWritten = true
+		}
+		wantAllAfterSuiteActions = false
+	})
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("StorkctlPerformFailoverFailbackDefaultMetroMultiple", "Failover and Failback using storkctl on metro cluster for multiple NS", nil, testrailID)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:496]: StartTorpedoTest has "StorkctlPerformFailoverFailbackDefaultMetroMultiple" but Describe has "StorkctlPerformFailoverFailbackPostgresql".

File: tests/basic/async_dr_test.go [hunk 3]
```diff
@@ -476,9 +481,172 @@ var _ = Describe("{StorkctlPerformFailoverFailbackDefaultMetroMultiple}", func()
...
+	BeforeEach(func() {
+		if !kubeConfigWritten {
+			// Write kubeconfig files after reading from the config maps created by torpedo deploy script
+			WriteKubeconfigToFiles()
+			kubeConfigWritten = true
+		}
+		wantAllAfterSuiteActions = false
+	})
+	JustBeforeEach(func() {
+		StartTorpedoTest("StorkctlPerformFailoverFailbackDefaultMetroMultiple", "Failover and Failback using storkctl on metro cluster for multiple NS", nil, testrailID)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:537]: StartTorpedoTest has "StorkctlPerformFailoverFailbackDefaultMetroMultiple" but Describe has "StorkctlPerformFailoverFailbackElasticSearch".

File: tests/basic/async_dr_test.go [hunk 4]
```diff
@@ -476,9 +481,172 @@ var _ = Describe("{StorkctlPerformFailoverFailbackDefaultMetroMultiple}", func()
...
+		if !kubeConfigWritten {
+			// Write kubeconfig files after reading from the config maps created by torpedo deploy script
+			WriteKubeconfigToFiles()
+			kubeConfigWritten = true
+		}
+		wantAllAfterSuiteActions = false
+	})
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("StorkctlPerformFailoverFailbackDefaultMetroMultiple", "Failover and Failback using storkctl on metro cluster for multiple NS", nil, testrailID)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:578]: StartTorpedoTest has "StorkctlPerformFailoverFailbackDefaultMetroMultiple" but Describe has "StorkctlPerformFailoverFailbackPostgresqlClusterwide".

File: tests/basic/async_dr_test.go [hunk 5]
```diff
@@ -476,9 +481,172 @@ var _ = Describe("{StorkctlPerformFailoverFailbackDefaultMetroMultiple}", func()
...
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:619]: StartTorpedoTest has "StorkctlPerformFailoverFailbackDefaultMetroMultiple" but Describe has "StorkctlPerformFailoverFailbackeckEsClusterwide".

File: tests/basic/async_dr_test.go [hunk 6]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+	extraArgs := map[string]string{
+		"namespaces":             namespace,
+		"kubeconfig":             kubeConfigPathSrc,
+		"exclude-resource-types": "ClusterServiceVersion,operatorconditions,OperatorGroup,InstallPlan,Subscription",
+		"exclude-selectors":      "olm.managed=true",
+	}
+	extraArgsFailoverFailback := map[string]string{
+		"kubeconfig": kubeConfigPathDest,
+	}
+	err = patchStashStrategy(crName)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1019]: I think the err is not handled.

File: tests/basic/async_dr_test.go [hunk 7]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+	if err != nil {
+		log.Errorf("Error getting podlist: %v", err)
+	}
+	err = asyncdr.WaitForPodToBeRunning(podListDest)
+	if err != nil {
+		log.Errorf("Error waiting for pod to be running: %v", err)
+	}
+	podCountDest := len(podListDest.Items)
+	dash.VerifyFatal(podCountDest == podCount, true, "Pod count is not equal to expected")
+	err = patchStashStrategy(crName)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1042]: I think the err is not handled.

File: tests/basic/async_dr_test.go [hunk 8]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+}
+
+func createOperatorBasedApp(appPath, opPath, ns string, clusterwide bool) (*v1.PodList, error) {
+	nsSpec := &v1.Namespace{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: ns,
+		},
+	}
+	_, err := core.Instance().CreateNamespace(nsSpec)
+	log.FailOnError(err, "Failed to create namespace %v on source", ns)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1118]: return err?

File: tests/basic/async_dr_test.go [hunk 9]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+		deployAndWaitForRunning(opPath, ns, "src")
+	}
+	deployAndWaitForRunning(appPath, ns, "src")
+	podList, err := core.Instance().GetPods(ns, nil)
+	if err != nil {
+		log.Errorf("Error getting podlist: %v and err is: %v", appPath, err)
+		return nil, err
+	}
+	err = SetDestinationKubeConfig()
+	log.FailOnError(err, "Failed to set destination kubeconfig")
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1131]: return err?

File: tests/basic/async_dr_test.go [hunk 10]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+	_, err = core.Instance().CreateNamespace(nsSpec)
+	log.FailOnError(err, "Failed to create namespace %v on destination", ns)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1133]: return err?

File: tests/basic/async_dr_test.go [hunk 11]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+		deployAndWaitForRunning(opPath, ns, "dest")
+	}
+	return podList, nil
+}
+
+func deployAndWaitForRunning(path, namespace, cluster string) error {
+	cmd := fmt.Sprintf("kubectl apply -f %v -n %v", path, namespace)
+	if cluster == "dest" {
+		kubeConfigPathDest, err := GetCustomClusterConfigPath(asyncdr.SecondCluster)
+		log.FailOnError(err, "Failed to get destination configPath: %v", err)
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1147]: return err?

File: tests/basic/async_dr_test.go [hunk 12]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+		cmd = fmt.Sprintf("kubectl apply -f %v -n %v --kubeconfig %v", path, namespace, kubeConfigPathDest)
+		err = hardSetConfig(kubeConfigPathDest)
+		log.FailOnError(err, "Failed to set destination kubeconfig")
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1150]: return err?

File: tests/basic/async_dr_test.go [hunk 13]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+	if err != nil {
+		log.Errorf("Error getting podlist: %v and err is: %v", cmd, err)
+		return err
+	}
+	err = asyncdr.WaitForPodToBeRunning(podList)
+	if err != nil {
+		return err
+	}
+	err = SetSourceKubeConfig()
+	log.FailOnError(err, "Failed to set source kubeconfig")
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1170]: return err?

File: tests/basic/async_dr_test.go [hunk 14]
```diff
@@ -817,10 +985,252 @@ func hardSetConfig(configPath string) error {
...
+		log.Infof("Error unmarshalling JSON: %v", err)
+		return 0, err
+	}
+	return len(appReg.Resources), nil
+}
+
+func patchStashStrategy(crName string) error {
+	var patches []PatchOperation
+	numResources, err := getNumResources(crName)
+	return err
```
Comment (kphalgun-px) [inline tests/basic/async_dr_test.go:1206]: How about?

```go
if err != nil {
    return errr
}
```

File: tests/basic/async_dr_test.go (hunk 2 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 3 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 4 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 5 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 6 above)
Comment (gejain) [reply tests/basic/async_dr_test.go:1019]: Done

File: tests/basic/async_dr_test.go (hunk 7 above)
Comment (gejain) [reply tests/basic/async_dr_test.go:1043]: Done

File: tests/basic/async_dr_test.go (hunk 8 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 9 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 10 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 11 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 12 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 13 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done

File: tests/basic/async_dr_test.go (hunk 14 above)
Comment (gejain) [reply tests/basic/async_dr_test.go]: Done, I missed it. Thanks!

**Review by kphalgun-px — APPROVED** [review]
LGTM

**Pull request merged**